python/src/lib/workbook_validator.py

```diff
@@ -97,38 +103,45 @@ def validate(workbook: IO[bytes]) -> Tuple[Errors, Optional[str]]:
     See: https://openpyxl.readthedocs.io/en/stable/optimized.html
     If there is any trouble loading files from memory please see here: https://stackoverflow.com/questions/20635778/using-openpyxl-to-read-file-from-memory
     """
-    wb = load_workbook(filename=workbook, read_only=True)
+    try:
+        wb = load_workbook(filename=workbook, read_only=True)
+        return validate_workbook(wb)
+    finally:
+        workbook.close()
 
+
+def validate_workbook(workbook: Workbook) -> Tuple[Errors, Optional[str]]:
+    """Validates a given Excel workbook according to CPF validation rules.
+
+    Args:
+        workbook: The Excel workbook to validate.
     """
-    2. Validate logic sheet to make sure the sheet has an appropriate version
-    """
-    errors = validate_logic_sheet(wb[LOGIC_SHEET])
-    if len(errors) > 0:
-        return errors, None
 
     """
-    3. Validate cover sheet and project selection. Pick the appropriate validator for the next step.
+    1. Validate logic sheet to make sure the sheet has an appropriate version
     """
-    errors, project_schema = validate_cover_sheet(wb[COVER_SHEET])
-    if len(errors) > 0:
-        return errors, None
-    
-    project_use_code = get_project_use_code(wb[COVER_SHEET])
+    errors: Errors = []
+    errors += validate_logic_sheet(workbook[LOGIC_SHEET])
 
     """
-    4. Ensure all project rows are validated with the schema
+    2. Validate cover sheet and project selection. Pick the appropriate validator for the next step.
     """
-    project_errors = validate_project_sheet(wb[PROJECT_SHEET], project_schema)
+    cover_errors, project_schema = validate_cover_sheet(workbook[COVER_SHEET])
+    errors += cover_errors
+
+    project_use_code = get_project_use_code(workbook[COVER_SHEET])
 
     """
-    5. Ensure all subrecipient rows are validated with the schema
+    3. Ensure all project rows are validated with the schema
     """
-    subrecipient_errors = validate_subrecipient_sheet(wb[SUBRECIPIENTS_SHEET])
+    errors += validate_project_sheet(workbook[PROJECT_SHEET], project_schema)
```

Comment:
Note: we should only validate this if `project_schema` is not `None` and I believe there is a small chance it could be.

Open to solving it in either way of these two ways here:
1. add a conditional to only execute this line _if_ there is a `project_schema`
```py
if project_schema:
    validate_project_sheet
```
or within `validate_project_sheet` function
```py
if !project_schema:
    return [WorkbookError("Unknown Project"]
```

Reply:
Good call